py/client/pydeephaven/session.py

```diff
@@ -352,24 +338,36 @@ def _connect(self):
 
             self.is_connected = True
 
-            self._timeout = int(session_duration.string_value)
+            self._timeout_seconds = int(session_duration.string_value)/1000.0
             if self._never_timeout:
                 self._keep_alive()
 
     def _keep_alive(self):
-        if self.is_connected:
-            if self._keep_alive_timer:
-                self._refresh_token()
-            self._keep_alive_timer = threading.Timer(self._timeout / 2 / 1000, self._keep_alive)
-            self._keep_alive_timer.daemon = True
-            self._keep_alive_timer.start()
-
-    def _refresh_token(self):
+        trace(f'_keep_alive')
+        if not self.is_connected:
+            return
+        factor = 1
+        ok = True
+        if self._keep_alive_timer:
+            ok = self._refresh_token()
+            if not ok:
+                factor = 0.5
+        timer_wakeup = self._timeout_seconds * factor
```

Comment:
Is this retry-logic necessary? I suppose it does make the whole keep-alive effort a bit more error tolerant in case of network hiccup and some other exceptional cases but it only makes a difference when the application is actually really idle for the moment.

Reply:
It is necessary, the alternative is terminating the client on an exception when a network glitch happens, which is unreasonable.